File: labyrinth.rb
# frozen_string_literal: true
require_relative 'directions'
require_relative 'orientation'
require_relative 'dice'
require_relative 'player'
require_relative 'monster'
module Irrgarten
  class Labyrinth
    @@BLOCK_CHAR = 'X'
    @@EMPTY_CHAR = '-'
    @@MONSTER_CHAR = 'M'
    @@COMBAT_CHAR = 'C'
    @@EXIT_CHAR = 'E'
    @@ROW = 0
    @@COL = 1

    def initialize(n_rows, n_cols, exit_row, exit_col)
      @n_rows = n_rows
      @n_cols = n_cols
      @exit_row = exit_row
      @exit_col = exit_col

      #si pongo Array.new(4, 5), me hace un array de 4 con cincos. por eso me sirve para inicializar con empty char
      @labyrinth = Array.new(@n_rows){Array.new(@n_cols, @@EMPTY_CHAR)}
      #@labyrinth = []
      #for i in 0...@n_rows
      #  @labyrinth[i] = []
      #  for j in 0...@n_cols
      #    @labyrinth[i][j] = @@EMPTY_CHAR
      #  end
      #end
      @players = Array.new(@n_rows){Array.new(@n_cols, nil)}
      @monsters = Array.new(@n_rows){Array.new(@n_cols, nil)}

      @labyrinth[@exit_row][@exit_col] = @@EXIT_CHAR
    end

    def spread_players(players)
      for player in players
        p = player
        pos = random_empty_pos
        put_player_2d(-1, -1, pos[@@ROW], pos[@@COL], p)
      end
    end

    def have_a_winner
      @players[@exit_row][@exit_col] != nil
    end

    def to_s
      cad = ""
      for i in 0...@labyrinth.size
        for j in 0...@labyrinth[i].size
          cad = cad + @labyrinth[i][j].to_s + " "
        end
        cad = cad + " \n"
      end
      cad
    end

    def add_monster(row, col, monster)
      if pos_OK(row,col) && empty_pos(row,col)
        @labyrinth[row][col] = @@MONSTER_CHAR
        @monsters[row][col] = monster
        #monster.pos(row,col)
      end
    end

    def put_player(directions, player)
      old_row = player.row
      old_col = player.col
      new_pos = dir_2_pos(old_row, old_col, directions)
      monster = put_player_2d(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
    end

    def add_block(orientation, start_row, start_col, length)
      if orientation == Orientation::VERTICAL
        inc_row = 1
        inc_col = 0
      else
        inc_row = 0
        inc_col = 1
      end

      row = start_row
      col = start_col

      while pos_OK(row, col) && empty_pos(row, col) && (length > 0)
        @labyrinth[row][col] = @@BLOCK_CHAR
        length -=1
        row += inc_row
        col += inc_col
      end
    end

    def valid_moves(row, col)
      output = []
      if can_step_on(row+1, col)
        output << (Directions::DOWN)
      end
      if can_step_on(row-1, col)
        output << (Directions::UP)
      end
      if can_step_on(row, col+1)
        output << (Directions::RIGHT)
      end
      if can_step_on(row, col-1)
        output << (Directions::LEFT)
      end
      output
    end

    private

    def pos_OK(row, col)
      if(row < @n_rows) && (row >= 0) && (col < @n_cols) && (col >= 0)
        true
      else
        false
      end
    end

    def empty_pos(row, col)
      @labyrinth[row][col] == @@EMPTY_CHAR
    end

    def monster_pos(row,col)
      @labyrinth[row][col] == @@MONSTER_CHAR
    end

    def exit_pos(row,col)
      @labyrinth[row][col] == @@EXIT_CHAR
    end

    def combat_pos(row,col)
      if (@labyrinth[row][col] == @@COMBAT_CHAR) && (@players[row][col] != nil)
        true
      else
        false
      end
    end

    def can_step_on(row,col)
      can = false
      if pos_OK(row,col)
        if empty_pos(row,col) || monster_pos(row,col) || exit_pos(row,col)
          can = true
        end
      end
      can
    end

    def update_old_pos(row,col)
      if pos_OK(row,col)
        if combat_pos(row,col)
          @labyrinth[row][col] = @@MONSTER_CHAR
        else
          @labyrinth[row][col] = @@EMPTY_CHAR
        end
      end
    end

    def dir_2_pos(row,col,direction)
      aux = [row,col]

      case direction
        when Directions::UP
          aux[@@ROW]=row-1
        when Directions::DOWN
          aux[@@ROW]=row+1
        when Directions::LEFT
          aux[@@COL]=col-1
        when Directions::RIGHT
          aux[@@COL]=col+1
      end
        aux
    end

    def random_empty_pos
      rand = Dice.new()
      vacia = false
      while !vacia
        pos = []
        pos.push(rand.random_pos(@n_rows))
        pos.push(rand.random_pos(@n_cols))

        if empty_pos(pos[0], pos[1])
          vacia = true
        end
      end
      pos
    end

    def put_player_2d(old_row, old_col, row, col, player)
      output = nil

      if can_step_on(row,col)
        if pos_OK(old_row,old_col)
          p = @players[old_row][old_col]
          if p == player
            update_old_pos(old_row,old_col)
            @players[old_row][old_col] = nil
          end
        end

        monster_pos = monster_pos(row, col)

        if monster_pos
          @labyrinth[row][col] = @@COMBAT_CHAR
          output = @monsters[row][col]
        else
          number = player.number
          @labyrinth[row][col] = number.to_s
        end

        @players[row][col] = player
        player.pos(row, col)
      end

      output
    end
  end
end

File: Labyrinth.java
/**
 *
 * @author gonzaalovd & carlitros_gamer20
 */

package irrgarten;

import java.util.ArrayList;

public class Labyrinth {

    static private final char BLOCK_CHAR = 'X';
    static private final char EMPTY_CHAR = '-';
    static private final char MONSTER_CHAR = 'M';       
    static private final char COMBAR_CHAR = 'C';
    static private final char COMBAT_CHAR = 'C';
    static private final char EXIT_CHAR = 'E';
    static private final int ROW = 0;
    static private final int COL = 1;

    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;

    //atributos de relación

    private Monster[][] monsters; //tamb se podría poner Monster mosters[][]
    private char[][] labyrinth;
    private Player[][] players;

    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;

        this.labyrinth = new char [this.nRows][this.nCols];
        this.players = new Player [this.nRows][this.nCols];
        this.monsters = new Monster [this.nRows][this.nCols];

        for(int i = 0; i < this.nRows; i++){
            for(int j = 0 ; j < this.nCols; j++){

                this.labyrinth[i][j] = Labyrinth.EMPTY_CHAR;
                this.players[i][j] = null;
                this.monsters[i][j] = null;

            }
        }

        this.labyrinth[this.exitRow][this.exitCol] = Labyrinth.EXIT_CHAR;
    }

    public boolean haveAWinner(){
        return this.players[this.exitRow][this.exitCol] != null;
    }

    public void spreadPlayers(ArrayList<Player> players){
        for (int i = 0; i<players.size(); i++){
            Player p = players.get(i);
            int pos[] = this.randomEmptyPos();
            this.putPlayer2D(-1, -1, pos[ROW], pos[COL], p);
        }
    }

    public String toString(){
        String str = "";
        
        for(int i = 0; i<this.nRows; i++){
            for(int j = 0; j<this.nCols; j++ ){
                str += this.labyrinth[i][j] + " ";
            }
            str += "\n";
        }
        
        return str;
    }

    public void addMonster(int row,int col, Monster monster){

        if(emptyPos(row,col) && posOK(row, col)){
            this.labyrinth[row][col] = Labyrinth.MONSTER_CHAR;
            this.monsters[row][col] = monster;
            monster.setPos(row, col);
        }

    }

    //public Monster putPLayer(Directions direction, Player player){
    public Monster putPlayer(Directions directions, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int newPos[] = this.dir2pos(oldRow, oldCol, directions);
        
        Monster monster = this.putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
        
        return monster;
    }

    //public void addBlock(Orientation orientation, int startRow,int startCol, int length){
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow, incCol;
        
        if(orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }else{
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;

        while(this.posOK(row, col) && this.emptyPos(row, col) && length > 0){
            this.labyrinth[row][col] = BLOCK_CHAR;
            length--;
            row += incRow;
            col += incCol;
        }
    }

    //public ArrayList<Directions> validMoves(int row,int col);
    public ArrayList<Directions> validMoves(int row,int col){
        ArrayList<Directions> output;
        
        output = new ArrayList<>();
        
        if(this.canStepOn(row+1, col))
            output.add(Directions.DOWN);
        if(this.canStepOn(row-1, col))
            output.add(Directions.UP);
        if(this.canStepOn(row, col+1))
            output.add(Directions.RIGHT);
        if(this.canStepOn(row, col-1))
            output.add(Directions.LEFT);
        
        return output;
    }

    private boolean posOK(int row, int col){
        return (row < this.nRows) && (row >= 0) && (col < this.nCols) && (col >= 0);
    }

    private boolean emptyPos(int row, int col){
        return this.labyrinth[row][col] == Labyrinth.EMPTY_CHAR;
    }

    private boolean monsterPos(int row, int col){
        return this.labyrinth[row][col] == Labyrinth.MONSTER_CHAR;
    }

    private boolean exitPos(int row, int col){
        return this.labyrinth[row][col] == Labyrinth.EXIT_CHAR;
    }

    private boolean combatPos(int row, int col){
        return (this.labyrinth[row][col] == Labyrinth.COMBAR_CHAR) && (this.players[row][col]!= null);
    }

    private boolean canStepOn(int row, int col){
        
        boolean can = false;
        if(posOK(row,col)){
            if(emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col)){
                can = true;
            }
        }
        return can;
        
    }

    private void updateOldPos(int row, int col){
        if(this.posOK(row, col)){
            if(this.combatPos(row,col)){
                this.labyrinth[row][col] = Labyrinth.MONSTER_CHAR;
            }else{
                this.labyrinth[row][col] = Labyrinth.EMPTY_CHAR;
            }
        }
    }

    public int[] dir2pos(int row, int col, Directions direction){

        int aux[]={row,col};
        if(direction == Directions.DOWN)
            aux[0]=row+1;
        if(direction == Directions.UP)
            aux[0]=row-1;
        if(direction == Directions.LEFT)
            aux[1]=col-1;
        if(direction == Directions.RIGHT)
            aux[1]=col+1;

        return aux;
    }

    private int [] randomEmptyPos(){
        int col = Dice.randomPos(this.nCols);
        int row = Dice.randomPos(this.nRows);

        while(!this.emptyPos(row, col)){

            col = Dice.randomPos(this.nCols);
            row = Dice.randomPos(this.nRows);
        }
        
        int [] pos =  new int[2];
        pos[0] = row;
        pos[1] = col;

        return pos;
    }

    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
        
        Monster output = null;

        if(this.canStepOn(row, col)){
            if(this.posOK(oldRow, oldCol)){
                Player p = this.players[oldRow][oldCol];
                if(p == player){
                    this.updateOldPos(oldRow, oldCol);
                    this.players[oldRow][oldCol]=null;
                }
            }
            
            boolean monsterPos = this.monsterPos(row, col);
            if(monsterPos){
                this.labyrinth[row][col] = COMBAT_CHAR;
                output = this.monsters[row][col];
            }else{
                char number = player.getNumber();
                this.labyrinth[row][col] = number;
            }
            
            this.players[row][col] = player;
            player.setPos(row, col);
        }
        
        return output;
    }
    
    public void setFuzzyPlayer(FuzzyPlayer fp){
        players[fp.getRow()][fp.getCol()] = fp;
    }
}
